pallets/name-service/src/lib.rs

```diff
@@ -0,0 +1,445 @@
+
+        //check if name is registered
+        ensure!(
+            UsernameRecords::<T>::contains_key(&username),
+            Error::<T>::NotRegistered
+        );
+        // check if its owned (can unwrap because of previous checks for the key)
+        ensure!(
+            UsernameRecords::<T>::get(&username).unwrap() == registrant,
+            Error::<T>::NotOwned
+        );
+
+        // check if we already have a primary
+        if PrimaryRecords::<T>::contains_key(registrant) {
+            PrimaryRecords::<T>::mutate(registrant, |old_username| {
+                *old_username = Some(username.clone())
+            });
+        } else {
+            PrimaryRecords::<T>::insert(registrant, &username);
+        }
+
+        Self::deposit_event(Event::NameSetAsPrimary {
+            owner: registrant,
+            username,
+        });
+        Ok(())
+    }
+
+    fn try_cancel_pending_register(
+        username: UserName,
+        registrant: ZkAddressType,
+    ) -> DispatchResult {
+        // Username checks
+        username_validation(&username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        let (hash_user, hash_address) = (
+            T::Hashing::hash_of(&username),
+            T::Hashing::hash_of(&registrant),
+        );
+
+        let pending_register_data =
+            PendingRegister::<T>::get(hash_user).ok_or(Error::<T>::UsernameNotFound)?;
+
+        ensure!(
+            pending_register_data.0 == hash_address,
+            Error::<T>::NotOwned
+        );
+
+        PendingRegister::<T>::remove(hash_user);
+
+        Self::deposit_event(Event::RegisterCanceled {
+            hash_username: hash_user,
+            hash_owner: hash_address,
+        });
+        Ok(())
+    }
+
+    fn try_remove_register(username: UserName, registrant: ZkAddressType) -> DispatchResult {
+        // Username checks
+        username_validation(&username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        ensure!(
+            UsernameRecords::<T>::contains_key(&username),
+            Error::<T>::NotRegistered
+        );
+
+        ensure!(
+            UsernameRecords::<T>::get(&username).unwrap() == registrant,
+            Error::<T>::NotOwned
+        );
+
+        UsernameRecords::<T>::remove(&username);
+
+        // check if the name we are removing is a primary name to keep storage synced
+        if let Ok(primary_username) = PrimaryRecords::<T>::try_get(registrant) {
+            if primary_username == username {
+                PrimaryRecords::<T>::remove(registrant);
+            }
+        }
+
+        Self::deposit_event(Event::RegisterRemoved {
+            username,
+            owner: registrant,
+        });
+        Ok(())
+    }
+}
+
+/// username validation
+fn username_validation(username: &Vec<u8>) -> Option<()> {
+    let label = core::str::from_utf8(username.as_slice())
+        .map(|label| label.to_ascii_lowercase())
```

Comment:
Why change to lowerecase? 

The above callls still store the uppercase version in storage

Reply:
Yeah will try to get a better regex check, was reusing some old checks from Quihuans draft. Bad idea i know